python_modules/dagster/dagster/_core/definitions/data_version.py

```diff
@@ -474,10 +476,38 @@ def _is_dep_updated(self, provenance: DataProvenance, dep_key: "AssetKeyPartitio
                     else None
                 )
                 updated_version = extract_data_version_from_entry(updated_record.event_log_entry)
-                return previous_version != updated_version
+                return self._data_versions_differ(previous_version, updated_version)
             else:
                 return False
 
+    def _data_versions_differ(
+        self, prev_data_version: Optional[DataVersion], curr_data_version: Optional[DataVersion]
+    ) -> bool:
+        # We special case this to handle a complex niche scenario:
+        #
+        # - Data version system uses
+        #     - INITIAL for assets that are assumed to have a value despite no recorded data version
+        #       (external assets are always assumed to have a value)
+        #     - NULL for assets that are assumed to have no value (dagster-managed assets that have
+        #       never been materialized)
+        # - Suppose we have an asset X and two code locs, A and B. Asset X is defined as a
+        #   materializable asset in A and external asset in B.
+        # - Because there is no record for the asset, the framework generates a default data version.
+        #     - In A, this will be NULL (because it is a dagster-managed asset with no materialization)
+        #     - In B, this will be INITIAL (because it is an external asset assumed to have a value)
+        # - This discrepancy is fine until you introduce another asset Y that depends on X in code
+        #   location B. When you materialize Y (in code loc B), it will save X's data version as
+        #   INITIAL in the data provenance
+        # - Now when we compare the version of X on Y provenance (INITIAL) to the current version in
+        #   this code (NULL), they will look different, though X was never changed.
+        #
+        # To mitigate this, we treat the case where the previous data version is
+        # DEFAULT_DATA_VERSION and current data version is NULL_DATA_VERSION as unchanged.
+        if prev_data_version == DEFAULT_DATA_VERSION and curr_data_version == NULL_DATA_VERSION:
```

Comment:
is there a case where you would have the reverse? NULL transitioning to DEFAULT?